pkg/tcpip/transport/tcp/endpoint.go

```diff
@@ -953,6 +953,12 @@ func (e *endpoint) Readiness(mask waiter.EventMask) waiter.EventMask {
 		}
 	}
 	if e.EndpointState().connected() {
+		// Determine whether endpoint is half-closed with rcv shutdown
+		e.mu.Lock()
```

Comment:
Readiness cannot acquire e.mu. This leads to a deadlock in epoll. I would suggest creating a flags field in the endpoint that is read atomically and use that instead of RcvClosed/SndClosed to identify when to deliver RdHup.

Reply:
Good point. I missed the endpoint struct Godoc that mentioned exactly that. Done.